akvo/rsr/project_role_auth.py

```diff
@@ -0,0 +1,78 @@
+from rules.permissions import permissions
+
+from akvo.rsr.models import ProjectRole, ProjectHierarchy
+from akvo.rest.viewsets import PublicProjectViewSet
+from akvo.rsr.permissions import (
+    GROUP_NAME_ADMINS, GROUP_NAME_ME_MANAGERS, GROUP_NAME_PROJECT_EDITORS, GROUP_NAME_USERS,
+    GROUP_NAME_ENUMERATORS, GROUP_NAME_USER_MANAGERS)
+
+
+class ProjectRolePermissionBackend(object):
+    def authenticate(self, username, password):
+        return None
+
+    def has_perm(self, user, perm, obj):
+        if not user.is_authenticated():
+            return False
+
+        groups = groups_from_permission(perm)
+        all_roles = ProjectRole.objects.filter(user=user, group__name__in=groups)
+        # obj=None calls are made by the DRF when checking if permissions
+        # should be given for a "view", and not a specific object.
+        if obj is None:
+            # Check if the user has a project role with specified groups that
+            # have the required permission, irrespective of which project the
+            # role exists for.
+            return all_roles.exists()
+
+        project = PublicProjectViewSet.get_project(obj)
+        if project is None:
+            return False
+
+        reporting_org = project.reporting_org
+        ancestor = project.ancestor()
+        try:
+            hierarchy_org = ProjectHierarchy.objects.get(root_project=ancestor).organisation
+        except ProjectHierarchy.DoesNotExist:
+            hierarchy_org = None
+
+        if (reporting_org is None or not reporting_org.use_project_roles) and (
+            hierarchy_org is None or not hierarchy_org.use_project_roles
+        ):
+            # If the organisation is not using project roles, let the default
+            # ObjectPermissionBackend handle the permisisons!
+            return False
+
+        if user.admin_of(reporting_org):
+            return True
+
+        roles = all_roles.filter(project=project)
+        if roles.exists():
+            return True
+        else:
+            raise PermissionDenied
+
+
+def groups_from_permission(permission):
+    # FIXME: Trying to use the rules definitions is clunky, but this can be
+    # cleaned up to make it easier to manage both project roles and
+    # employment based permissions.
+    predicate = permissions.get(permission, None)
+    if predicate is None:
+        return []
+
+    rules = predicate.name
+    # FIXME: We assume that only | is being used, and no &
+    rules = rules.replace('(', '').replace(')', '').replace('|', '').split()
+
+    rule_to_group_mapping = {
+        'is_org_admin': [GROUP_NAME_ADMINS],
+        'is_org_user_manager': [GROUP_NAME_USER_MANAGERS],
+        'is_org_me_manager_or_project_editor': [GROUP_NAME_PROJECT_EDITORS, GROUP_NAME_ME_MANAGERS],
+        'is_org_me_manager': [GROUP_NAME_ME_MANAGERS],
+        'is_org_user': [GROUP_NAME_USERS],
+        'is_org_enumerator': [GROUP_NAME_ENUMERATORS],
+    }
```

Comment:
I'm not really sure about this, it's a kind of duplication with predicate definitions, _i.e._ we need to maintain it in two different places.

Is it possible if the project role permission logics are implemented as `rules.predicate` instead of auth backend? I think that will be more future-proof and can eliminate the problem of parsing predicate names and  assumes only `|` operator will be used.

And I've just noticed that we are using a very old version of `django-rules`. but I don't think that's an important issue right now.

Reply:
You are correct about the duplications, but we could get rid of the duplications and maintain the same list of groups to permission mapping that is used both by the project permissions layer and the organisation employment permissions layer. 

I see that there would be some gains to use a `rules.predicate` for the project level permissions, but I don't really like the fact that each predicate is run separately for each rule, and the org level rules are sometimes quite slow to compute. Also, we'd need to make sure we correctly add project level predicates wherever we have org level predicates, and make sure we keep doing that in the future too. 

The other option could be to combine the project level checks into the org level predicates, and call them something more generic. But, still this would mean the predicates are called individually, and each of the checks is run multiple times inside each predicate. 

Having the project level permissions as a separate layer in front of the organisation permissions layer allows us to be able to just run through one function and decide if a user has any project level permissions or not.